File: axo_vem/axo_vem/infrastructure/transport/api/controllers/jobs.py
from __future__ import annotations

from typing import Any, Callable, Dict, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from pymongo.collection import Collection
from xolo.client.models import UserDTO

from axo_shared import wire
from axo_shared.protocol import Command

from axo_vem.application.compute.submit_function_job import SubmitFunctionJobUseCase
from axo_vem.domain.errors import NotFoundError
from axo_vem.domain.execution.repository import JobRepository
from axo_vem.infrastructure.transport.zmq_command.endpoint_client import resolve_rpc_uri, send_command


class _JobSubmitRequest(BaseModel):
    function_id: str
    function_name: Optional[str] = None
    function_version: int
    params: Dict[str, Any] = {}


class _FunctionJobSubmitRequest(BaseModel):
    params: Dict[str, Any] = {}


def build_router(
    job_repository: JobRepository,
    current_user_dependency: Callable[..., UserDTO],
    endpoints: Optional[Collection] = None,
    command_timeout_seconds: float = 3.0,
    submit_function_job_use_case: Optional[SubmitFunctionJobUseCase] = None,
) -> APIRouter:
    router = APIRouter(tags=["job telemetry"])

    if submit_function_job_use_case is not None:

        @router.post("/functions/{function_id}/{version}/jobs", tags=["function registries"])
        def submit_function_job(
            function_id: str,
            version: int,
            body: _FunctionJobSubmitRequest,
            _current_user: UserDTO = Depends(current_user_dependency),
        ):
            """Callers never pick an endpoint -- the target node is resolved
            server-side from the function's own virtual_environment_id (see
            SubmitFunctionJobUseCase). Returns the same shape
            POST /endpoints/{endpoint_id}/jobs does, plus which endpoint_id
            it actually landed on, so the caller can poll
            GET /endpoints/{endpoint_id}/jobs/{job_id} for the result."""
            return submit_function_job_use_case.execute(
                function_id=function_id, version=version, params=body.params,
            )

    @router.get("/jobs/{job_id}")
    def get_job(job_id: str, _current_user: UserDTO = Depends(current_user_dependency)):
        job = job_repository.get(job_id)
        if job is None:
            raise NotFoundError("job not found")
        return job.to_dict()

    @router.get("/functions/{function_id}/jobs")
    def list_jobs_for_function(
        function_id: str,
        version: Optional[int] = None,
        _current_user: UserDTO = Depends(current_user_dependency),
    ):
        """Job history for one function (optionally scoped to one version),
        newest first -- backs the Jobs table on the function detail page.
        Sourced from the jobs read model, not a live node query."""
        return [job.to_dict() for job in job_repository.list_by_function(function_id, version)]

    if endpoints is not None:

        @router.post("/endpoints/{endpoint_id}/jobs", tags=["endpoint management"])
        def submit_job(
            endpoint_id: str,
            body: _JobSubmitRequest,
            _current_user: UserDTO = Depends(current_user_dependency),
        ):
            """Proxies JOB_SUBMIT straight to the target endpoint's ROUTER --
            same relay shape as register_function/create_bucket. Submission
            is fire-and-forget from this route's perspective (JOB_SUBMIT
            itself is queued/async on the node); poll
            GET /endpoints/{endpoint_id}/jobs/{job_id} for the result.

            function_id is the node's actual catalog lookup key (see
            JobSubmitHandler.handle()) -- unlike register_function/
            delete_function/update_function, which only ever get a caller-
            supplied name and must derive function_id themselves via
            compute_function_id, the caller here already knows function_id
            (it's what GET /functions/{function_id} was keyed on), so it's
            taken directly rather than re-derived. function_name, if given,
            travels along only for the node's own logging."""
            doc = endpoints.find_one({"_id": endpoint_id})
            if doc is None:
                raise HTTPException(status_code=404, detail="endpoint not found")
            router_bind = doc.get("router_bind")
            if not router_bind:
                raise HTTPException(status_code=409, detail="endpoint has no known router_bind")
            rpc_uri = resolve_rpc_uri(router_bind, endpoint_id)

            command = Command(
                operation=wire.JOB_SUBMIT,
                content_type="application/json",
                envelope={
                    "function_id": body.function_id,
                    "function_name": body.function_name,
                    "function_version": body.function_version,
                    "params": body.params,
                },
            )
            result = send_command(rpc_uri, command, command_timeout_seconds)
            if result.is_err:
                raise HTTPException(status_code=504, detail=str(result.unwrap_err()))
            command_result = result.unwrap()
            if not command_result.ok:
                raise HTTPException(status_code=409, detail=command_result.error)
            return {"endpoint_id": endpoint_id, **(command_result.metadata or {})}

        @router.get("/endpoints/{endpoint_id}/jobs/{job_id}", tags=["endpoint management"])
        def poll_job_result(
            endpoint_id: str,
            job_id: str,
            _current_user: UserDTO = Depends(current_user_dependency),
        ):
            """Relays JOB_RESULT directly to the node -- the low-latency
            polling path, ahead of the Kurrent-to-Mongo pipeline, and the
            only source of a job's actual result values/error today (the
            jobs read model deliberately doesn't persist them, only
            status/timing -- see domain/execution/job.py)."""
            doc = endpoints.find_one({"_id": endpoint_id})
            if doc is None:
                raise HTTPException(status_code=404, detail="endpoint not found")
            router_bind = doc.get("router_bind")
            if not router_bind:
                raise HTTPException(status_code=409, detail="endpoint has no known router_bind")
            rpc_uri = resolve_rpc_uri(router_bind, endpoint_id)

            command = Command(
                operation=wire.JOB_RESULT,
                content_type="application/json",
                envelope={"job_id": job_id},
            )
            result = send_command(rpc_uri, command, command_timeout_seconds)
            if result.is_err:
                raise HTTPException(status_code=504, detail=str(result.unwrap_err()))
            command_result = result.unwrap()
            if not command_result.ok:
                raise HTTPException(status_code=409, detail=command_result.error)
            return {"job_id": job_id, "endpoint_id": endpoint_id, **(command_result.metadata or {})}

    return router

File: axo-ui/app/stores/jobs.ts
import type { JobHistoryEntry, JobResult } from '~/types/job'

const TERMINAL_STATUSES = new Set(['COMPLETED', 'FAILED'])

export const useJobsStore = defineStore('jobs', () => {
  const { apiFetch } = useApi()

  const current = ref<JobResult | null>(null)
  const history = ref<JobHistoryEntry[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  // functionId is the node's actual catalog lookup key (the route param on
  // /functions/{id}/...), not a display name -- see JobSubmitHandler.handle()
  // on the node side. functionName is optional and only ever used for the
  // node's own logging.
  async function submit(
    endpointId: string,
    functionId: string,
    functionVersion: number,
    params: Record<string, unknown>,
    functionName?: string,
  ) {
    error.value = null
    try {
      return await apiFetch<JobResult>(`/endpoints/${endpointId}/jobs`, {
        method: 'POST',
        body: { function_id: functionId, function_name: functionName, function_version: functionVersion, params },
      })
    } catch (err: any) {
      error.value = err.message
      throw err
    }
  }

  // Caller never picks an endpoint -- the API resolves one server-side from
  // the function's own virtual_environment_id (preferring that VE's
  // currently-elected mesh leader). Returns the same JobResult shape submit()
  // does, plus which endpoint_id it actually landed on, needed for poll().
  async function submitForFunction(
    functionId: string,
    functionVersion: number,
    params: Record<string, unknown>,
  ) {
    error.value = null
    try {
      return await apiFetch<JobResult>(`/functions/${functionId}/${functionVersion}/jobs`, {
        method: 'POST',
        body: { params },
      })
    } catch (err: any) {
      error.value = err.message
      throw err
    }
  }

  async function poll(endpointId: string, jobId: string) {
    error.value = null
    try {
      current.value = await apiFetch<JobResult>(`/endpoints/${endpointId}/jobs/${jobId}`)
      return current.value
    } catch (err: any) {
      error.value = err.message
      throw err
    }
  }

  // Submit-then-poll loop until a terminal status or timeout -- mirrors
  // AxoEndpointClient.run()'s convenience wrapper on the Python client side.
  async function run(
    endpointId: string,
    functionId: string,
    functionVersion: number,
    params: Record<string, unknown>,
    { intervalMs = 800, timeoutMs = 30000 }: { intervalMs?: number, timeoutMs?: number } = {},
  ) {
    loading.value = true
    error.value = null
    try {
      const submitted = await submit(endpointId, functionId, functionVersion, params)
      const jobId = submitted.job_id
      const startedAt = Date.now()

      while (true) {
        const result = await poll(endpointId, jobId)
        if (result.status && TERMINAL_STATUSES.has(result.status)) return result
        if (Date.now() - startedAt > timeoutMs) throw new Error('Timed out waiting for the job to finish')
        await new Promise(resolve => setTimeout(resolve, intervalMs))
      }
    } catch (err: any) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  async function fetchHistory(functionId: string, version?: number) {
    error.value = null
    try {
      history.value = await apiFetch<JobHistoryEntry[]>(`/functions/${functionId}/jobs`, {
        params: version !== undefined ? { version } : undefined,
      })
    } catch (err: any) {
      error.value = err.message
    }
  }

  return { current, history, loading, error, submit, submitForFunction, poll, run, fetchHistory }
})
